### DISCUSSION TITLE
Best practice route middleware for checking access permission

### DISCUSSION THREAD

**Discussion by procrates** [nuxt/nuxt#17926] [Questions] [answered]
```
export default defineNuxtRouteMiddleware(async (to, from) => {

  const { data } = await useFetch("/api/auth/check-role", {
    headers: useRequestHeaders(['cookie'])
  })
  if (data.value !== "admin") {
    return navigateTo({ path: "/" })
  }
  return
})
```

Returns: 
```
[Vue warn]: onServerPrefetch is called when there is no active component instance to be associated with. Lifecycle injection APIs can only be used during execution of setup(). If you are using async setup(), make sure to register lifecycle hooks before the first await statement.
```

It works, but I have a feeling that it is the wrong way to do this. Could someone with the knowledge on this help me with an example?

Also is this a secure way to check if someone has permission to enter the route? Before the route middleware I was using server middleware with route before:hook to check it. 

Thanks! Awsome work!

**Comment by Diizzayy** [comment, marked as answer]
@procrates 

> It works, but I have a feeling that it is the wrong way to do this

The warning that you're receiving is because `useFetch` should only be used within setup or Lifecycle Hooks.
You need to replace `useFetch` with `$fetch` in your middleware.

```js
export default defineNuxtRouteMiddleware(async (to, from) => {
  const { data } = await $fetch("/api/auth/check-role", {
    headers: useRequestHeaders(["cookie"]),
  });

  if (data?.value !== "admin") {
    return navigateTo({ path: "/" });
  }
});

```

> Also is this a secure way to check if someone has permission to enter the route?

Indeed! It's secure

**Reply by procrates** [reply]
Cool! Thanks :D
